marshal/types_test.go

```diff
@@ -0,0 +1,129 @@
+package marshal
+
+import (
+	"encoding/json"
+	"fmt"
+	"reflect"
+	"testing"
+
+	"github.com/stretchr/testify/assert"
+	"golang.org/x/exp/slices"
+)
+
+type customStruct struct {
+	// Basic types
+	Int   int     `json:"int,omitempty"`
+	Float float64 `json:"float,omitempty"`
+	Str   string  `json:"string,omitempty"`
+	Bool  bool    `json:"bool,omitempty"`
+
+	// Pointers to basic types
+	PInt   *int     `json:"pint,omitempty"`
+	PFloat *float64 `json:"pfloat,omitempty"`
+	PStr   *string  `json:"pstring,omitempty"`
+	PBool  *bool    `json:"pbool,omitempty"`
+
+	// Other types and structs
+	Slice      []int                    `json:"slice,omitempty"`
+	Interface  interface{}              `json:"interface,omitempty"`
+	BasicMap   map[string]string        `json:"basicmap,omitempty"`
+	MapToArray map[string][]interface{} `json:"maptoarray,omitempty"`
+	MapToInt   map[string]interface{}   `json:"maptoint,omitempty"`
+
+	ChildFS   structFS   `json:"childfs,omitempty"`
+	ChildNoFS structNoFS `json:"childnofs,omitempty"`
+
+	PChildFS   *structFS   `json:"pchildfs,omitempty"`
+	PChildNoFS *structNoFS `json:"pchildnofs,omitempty"`
+
+	MapToCustomTypeFS map[string]structFS `json:"maptocustomtypefs,omitempty"`
+
+	ForceSendFields []string `json:"-"`
+}
+
+type structNoFS struct {
+	Int int `json:"childint,omitempty"`
+}
+
+type structFS struct {
+	Int int `json:"childint,omitempty"`
+
+	ForceSendFields []string `json:"-"`
+}
+
+func (s *structFS) UnmarshalJSON(b []byte) error {
+	return Unmarshal(b, s)
+}
+
+func (s structFS) MarshalJSON() ([]byte, error) {
+	return Marshal(s)
+}
+
+func (s *customStruct) UnmarshalJSON(b []byte) error {
+	return Unmarshal(b, s)
+}
+
+func (s customStruct) MarshalJSON() ([]byte, error) {
+	return Marshal(s)
+}
+
+func Ptr[T any](v T) *T {
+	return &v
+}
+
+type basicMarshalTest struct {
+	st         customStruct
+	jsonString string
+	// Compare marshal results with normal marshal
+	matchClassic bool
+	// Unmarshal may not match, since ForceSendFields will be populated during
+	// custom unmarshal process
+	matchUnmarshal          bool
```

Comment:
Let's change this to the target unmarshal'ed datastructure. As a shorthand, we can use `nil` to mean `we expect st`.

Reply:
Changed